tools/animation/easing-visualizer.html

```diff
@@ -0,0 +1,1428 @@
+
+    // ==================== INPUT HANDLING ====================
+    function updateInputs() {
+      inputX1.value = state.bezier[0].toFixed(2);
+      inputY1.value = state.bezier[1].toFixed(2);
+      inputX2.value = state.bezier[2].toFixed(2);
+      inputY2.value = state.bezier[3].toFixed(2);
+    }
+
+    function handleInputChange() {
+      var x1 = Math.max(0, Math.min(1, parseFloat(inputX1.value) || 0));
+      var y1 = Math.max(-2, Math.min(2, parseFloat(inputY1.value) || 0));
+      var x2 = Math.max(0, Math.min(1, parseFloat(inputX2.value) || 0));
+      var y2 = Math.max(-2, Math.min(2, parseFloat(inputY2.value) || 0));
+      
+      state.bezier = [x1, y1, x2, y2];
+      state.currentPreset = findMatchingPreset() || 'custom';
+      
+      drawCurve();
+      updateCanvasInfo();
+      updateExport();
+      updatePresetButtons();
+      updateAnimationTiming();
+      saveToUrl();
+    }
+
+    function findMatchingPreset() {
+      var b = state.bezier;
+      var found = EASING_PRESETS.find(function(p) {
+        return Math.abs(p.bezier[0] - b[0]) < 0.01 &&
+               Math.abs(p.bezier[1] - b[1]) < 0.01 &&
+               Math.abs(p.bezier[2] - b[2]) < 0.01 &&
+               Math.abs(p.bezier[3] - b[3]) < 0.01;
+      });
+      return found ? found.id : null;
+    }
+
+    [inputX1, inputY1, inputX2, inputY2].forEach(function(input) {
+      input.addEventListener('input', handleInputChange);
+    });
+
+    // ==================== CANVAS DRAG ====================
+    function getMousePos(e) {
+      var rect = canvas.getBoundingClientRect();
+      return { x: e.clientX - rect.left, y: e.clientY - rect.top };
+    }
+
+    function hitTest(pos, pointX, pointY) {
+      var px = toCanvasX(pointX);
+      var py = toCanvasY(pointY);
+      var dx = pos.x - px;
+      var dy = pos.y - py;
+      return Math.sqrt(dx * dx + dy * dy) < 15;
+    }
+
+    canvas.addEventListener('mousedown', function(e) {
+      var pos = getMousePos(e);
+      if (hitTest(pos, state.bezier[0], state.bezier[1])) dragging = 'p1';
+      else if (hitTest(pos, state.bezier[2], state.bezier[3])) dragging = 'p2';
+    });
+
+    canvas.addEventListener('mousemove', function(e) {
+      if (!dragging) return;
+      var pos = getMousePos(e);
+      var x = Math.max(0, Math.min(1, fromCanvasX(pos.x)));
+      var y = Math.max(-0.5, Math.min(1.5, fromCanvasY(pos.y)));
+      
+      if (dragging === 'p1') {
+        state.bezier[0] = x;
+        state.bezier[1] = y;
+      } else {
+        state.bezier[2] = x;
+        state.bezier[3] = y;
+      }
+      
+      state.currentPreset = findMatchingPreset() || 'custom';
+      drawCurve();
+      updateInputs();
+      updateCanvasInfo();
+      updateExport();
+      updatePresetButtons();
+      updateAnimationTiming();
+    });
+
+    canvas.addEventListener('mouseup', function() { dragging = null; saveToUrl(); });
+    canvas.addEventListener('mouseleave', function() { dragging = null; });
```

Comment:
`mouseleave` handler does not call `saveToUrl()` while `mouseup` (line 1084) and `touchend` (line 1109) both do. When user drags a control point and moves mouse outside the canvas, the URL state is not saved.

```suggestion
    canvas.addEventListener('mouseleave', function() { dragging = null; saveToUrl(); });
```